agent/exec/plugin/controller.go

```diff
@@ -0,0 +1,148 @@
+package plugin
+
+import (
+	engineapi "github.com/docker/docker/client"
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// controller implements agent.Controller against docker's API.
+//
+// Most operations against docker's API are done through the plugin name,
+// which is unique to the task.
+type controller struct {
+	task    *api.Task
+	adapter *pluginAdapter
+	closed  chan struct{}
+	err     error
+}
+
+var _ exec.Controller = &controller{}
+
+// NewController returns a docker exec controller for the provided task.
+func NewController(client engineapi.APIClient, task *api.Task) (exec.Controller, error) {
+	adapter, err := newPluginAdapter(client, task)
+	if err != nil {
+		return nil, err
+	}
+
+	return &controller{
+		task:    task,
+		adapter: adapter,
+		closed:  make(chan struct{}),
+	}, nil
+}
+
+func (r *controller) Task() (*api.Task, error) {
```

Comment:
`c` is probably okay. `r` is leftover from when this `controller` had a different name.